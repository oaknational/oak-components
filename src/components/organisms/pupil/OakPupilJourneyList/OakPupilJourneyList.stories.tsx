import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakPupilJourneyList } from "./OakPupilJourneyList";

import { OakPupilJourneyListItem } from "@/components/organisms/pupil/OakPupilJourneyListItem";
import { OakPupilJourneyHeader } from "@/components/organisms/pupil/OakPupilJourneyHeader";

const meta: Meta<typeof OakPupilJourneyList> = {
  component: OakPupilJourneyList,
  tags: ["autodocs"],
  title: "components/organisms/pupil/OakPupilJourneyList",
  argTypes: {
    phase: { control: { options: ["primary", "secondary"] } },
  },
  parameters: {
    controls: {
      include: ["phase", "titleSlot", "counterSlot"],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakPupilJourneyList>;

export const Default: Story = {
  render: (args) => (
    <OakPupilJourneyList
      {...args}
      titleSlot={
        <OakPupilJourneyHeader
          title="Primary Lessons"
          iconName="subject-maths"
          breadcrumbs={["first", "second", "third"]}
        />
      }
      counterSlot={<div>Counter Slot Here</div>}
    >
      <OakPupilJourneyListItem title="Lesson 1" index={1} href="#" />
      <OakPupilJourneyListItem title="Lesson 2" index={2} href="#" />
      <OakPupilJourneyListItem title="Lesson 3" index={3} href="#" />
    </OakPupilJourneyList>
  ),

  args: {
    phase: "primary",
  },
};

export const NoTitle: Story = {
  render: (args) => (
    <OakPupilJourneyList {...args} counterSlot={<div>Counter Slot Here</div>}>
      <OakPupilJourneyListItem title="Lesson 1" index={1} href="#" />
      <OakPupilJourneyListItem title="Lesson 2" index={2} href="#" />
      <OakPupilJourneyListItem title="Lesson 3" index={3} href="#" />
    </OakPupilJourneyList>
  ),

  args: {
    phase: "primary",
  },
};
