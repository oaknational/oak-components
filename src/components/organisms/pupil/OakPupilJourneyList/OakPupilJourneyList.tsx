import React from "react";

import { OakFlex } from "@/components/atoms/OakFlex";
import { OakHandDrawnHR } from "@/components/molecules/OakHandDrawnHR";

export type OakPupilJourneyListProps = {
  children: React.ReactNode;
  phase: "primary" | "secondary";
  titleSlot?: React.ReactNode;
  counterSlot: React.ReactNode;
};

/**
 *
 * A styled list container for use with OakPupilJourneyListItems
 *
 *
 */

export const OakPupilJourneyList = ({
  children,
  phase,
  titleSlot,
  counterSlot,
}: OakPupilJourneyListProps) => {
  const outerBackgroundColor =
    phase === "primary"
      ? "bg-decorative4-very-subdued"
      : "bg-decorative3-very-subdued";
  const backgroundColor =
    phase === "primary" ? "bg-decorative4-subdued" : "bg-decorative3-subdued";
  return (
    <OakFlex
      $flexDirection={"column"}
      $width={["100%", "all-spacing-22", "all-spacing-23"]}
      $background={outerBackgroundColor}
    >
      {titleSlot && (
        <OakFlex
          $flexDirection={"column"}
          $gap={["space-between-m", "space-between-m2", "space-between-m2"]}
        >
          {titleSlot}
          <OakHandDrawnHR hrColor={"white"} $height={"all-spacing-1"} />
        </OakFlex>
      )}
      <OakFlex $mv={"space-between-m"}>{counterSlot}</OakFlex>

      <OakFlex
        $flexDirection={"column"}
        $pa={"inner-padding-m"}
        $borderRadius={"border-radius-l"}
        $gap={"space-between-s"}
        $background={backgroundColor}
        role="list"
      >
        {children}
      </OakFlex>
    </OakFlex>
  );
};
