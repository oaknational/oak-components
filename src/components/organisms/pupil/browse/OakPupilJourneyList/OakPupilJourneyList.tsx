import React from "react";

import { OakFlex } from "@/components/atoms/OakFlex";
import { OakHandDrawnHR } from "@/components/molecules/OakHandDrawnHR";

export type OakPupilJourneyListProps = {
  children: React.ReactNode;
  phase: "primary" | "secondary";
  titleSlot?: React.ReactNode;
  filterSlot?: React.ReactNode;
  subheadingSlot: React.ReactNode;
};

/*
 
FIXME: 
This component falls into the swiss army knife trap. 
Conditional logic is being used to handle a range of specific layout needs at a higher level in the tree.
It would be better to decompose this component and recompose more specific components handling the specific layout needs.
However, I do not want to introduce break changes in this PR.

*/

const Slots = ({
  titleSlot,
  filterSlot,
  subheadingSlot,
}: Pick<
  OakPupilJourneyListProps,
  "titleSlot" | "filterSlot" | "subheadingSlot"
>) => {
  if (titleSlot) {
    return (
      <OakFlex $flexDirection={"column"} $gap={"spacing-24"}>
        <OakFlex $flexDirection={"column"} $gap={"spacing-32"}>
          {titleSlot}
          <OakHandDrawnHR hrColor={"white"} $height={"spacing-4"} />
        </OakFlex>
        <OakFlex $flexDirection={"column"} $gap={"spacing-32"}>
          {filterSlot}
          <OakFlex>{subheadingSlot}</OakFlex>
        </OakFlex>
      </OakFlex>
    );
  } else {
    return <OakFlex $pt={"spacing-24"}>{subheadingSlot}</OakFlex>;
  }
};

/**
 *
 * A styled list container for use with OakPupilJourneyListItems
 *
 *
 */

export const OakPupilJourneyList = ({
  children,
  phase,
  titleSlot,
  subheadingSlot,
  filterSlot,
}: OakPupilJourneyListProps) => {
  const outerBackgroundColor =
    phase === "primary"
      ? "bg-decorative4-very-subdued"
      : "bg-decorative3-very-subdued";
  const backgroundColor =
    phase === "primary" ? "bg-decorative4-subdued" : "bg-decorative3-subdued";

  return (
    <OakFlex
      $flexDirection={"column"}
      $width={["100%", "spacing-640", "spacing-960"]}
      $background={outerBackgroundColor}
      $gap={"spacing-24"}
    >
      <Slots
        titleSlot={titleSlot}
        filterSlot={filterSlot}
        subheadingSlot={subheadingSlot}
      />
      <OakFlex
        $flexDirection={"column"}
        $pa={"spacing-16"}
        $borderRadius={"border-radius-l"}
        $gap={"spacing-16"}
        $background={backgroundColor}
        role="list"
      >
        {children}
      </OakFlex>
    </OakFlex>
  );
};
