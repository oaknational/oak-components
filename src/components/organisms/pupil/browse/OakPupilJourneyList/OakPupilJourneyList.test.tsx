import React from "react";
import "@testing-library/jest-dom";

import { OakPupilJourneyList } from "./OakPupilJourneyList";

import renderWithTheme from "@/test-helpers/renderWithTheme";

describe("OakPupilJourneyList", () => {
  it("should render successfully", () => {
    const { getByText } = renderWithTheme(
      <OakPupilJourneyList
        phase="primary"
        subheadingSlot={<div>Subheading Slot Here</div>}
      >
        <p>Hello World</p>
      </OakPupilJourneyList>,
    );
    expect(getByText("Hello World")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <OakPupilJourneyList
        phase="primary"
        subheadingSlot={<div>Subheading Slot Here</div>}
      >
        <p>Hello World</p>
      </OakPupilJourneyList>,
    );
    expect(container).toMatchSnapshot();
  });
});
