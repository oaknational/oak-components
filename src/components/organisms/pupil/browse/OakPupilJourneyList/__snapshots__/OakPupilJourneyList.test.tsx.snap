// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakPupilJourneyList matches snapshot 1`] = `
.c0 {
  width: 100%;
  background: #f5e9f2;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c2 {
  padding-top: 1.5rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c4 {
  padding: 1rem;
  background: #e5d1e0;
  border-radius: 1rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1.5rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1rem;
}

@media (min-width:750px) {
  .c0 {
    width: 40rem;
  }
}

@media (min-width:1280px) {
  .c0 {
    width: 60rem;
  }
}

<div
  className="c0 c1"
>
  <div
    className="c2 c3"
  >
    <div>
      Subheading Slot Here
    </div>
  </div>
  <div
    className="c4 c5"
    role="list"
  >
    <p>
      Hello World
    </p>
  </div>
</div>
`;
